node/db/src/rolling/mod.rs

```diff
@@ -0,0 +1,33 @@
+// Copyright 2019-2023 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+mod gc;
+pub use gc::*;
+mod impls;
+
+use std::{
+    collections::VecDeque,
+    path::{Path, PathBuf},
+    sync::Arc,
+};
+
+use forest_utils::db::file_backed_obj::FileBacked;
+use log::{info, warn};
+use parking_lot::RwLock;
+use serde::{Deserialize, Serialize};
+
+use crate::db_engine::{open_db, Db, DbConfig};
+
+#[derive(Clone)]
+pub struct RollingDB {
+    db_root: Arc<PathBuf>,
+    db_config: Arc<DbConfig>,
+    db_index: Arc<RwLock<FileBacked<DbIndex>>>,
+    /// A queue of active databases, from youngest to oldest
+    db_queue: Arc<RwLock<VecDeque<Db>>>,
```

Comment:
Good job making `db_queue` abstract.

Since we always want at least one DB, let's use a tuple for the queue: `db_queue: Arc<RwLock<(Db, VecDeque<Db>)>>`

Reply:
Could u elaborate on the benefit?   

I agree `(Db, VecDeque<Db>)` is a much better byte for public API. However, to implement this public API, I would still use `VecDeque<Db>` as internal implementation as it's a straightforward and direct mapping to the actual folder structure, making the code that maintains the relationship between the in-memory data structure and the actual underlining directories more straightforward and less error-prone.

In this case, `db_queue` is an internal field that is only accessible in the same module, I don't see why we should change it

Reply:
It's an invariant that there should always be at least 1 DB. We can statically enforce that invariant through the type-system.

The rest of the code doesn't have to care and can keep using a single `VecDeque<Db>`.

Later, I want to show that we always want 1 older Db. Then we can use `(Db, Db)`.

Reply:
fixed

Reply:
Changed the DB fields as requested, although I would still prefer decoupling DB design from the specific GC approach, designing DB in a GC agnostic way, and building it from bottom up

Reply:
Your objection has been noted.